### DISCUSSION TITLE
Writing docs as working examples

### DISCUSSION THREAD

**Discussion by pauleveritt** [Finistere/antidote#41] [General]
Hi, I am doing some work in a fork to write [fully working examples](https://antidote-examples.readthedocs.io/en/examples/examples/index.html) on things I'm interested in. If it turns out useful I'll do a PR.

I have a lot more I'd like to write. Which alas means, a lot of bugging @Finistere 😀

**Comment by Finistere** [comment]
I'm impressed by the quantity and the speed at which you wrote all of this. It would have taken me so much longer. ^^
I went through your examples and I get the following out of it:

- better integration `@dataclass` would be nice
- as discussed in the PR, `override` support for implementation
- scopes need to be reworked, I'm pretty sure it can be made easier or at least less error-prone. Currently their only function is to specify when a dependency value should be thrown out. If a dependency also depends on it, it generally must be marked with the same scope to handle it properly.
- for the `Injected Customer` section, you could inject it with something like this:

```python
class ForCurrentCustomer:
    @inject
    def evaluate(self,
                 predicate: Optional[QualifiedBy],
                 visit: Visit = inject.me(source=VisitHandler)) -> bool:
        if predicate is None:
            return False  # maybe?
        customer_type = visit.customer.__class__
        return any(customer_type == qualifier for qualifier in predicate.qualifiers)
```

If you need context, in the `@interface` paradigm, it should be injected either statically in the predicate `weight()` method or at runtime in the constraint's `evaluate()`.

In the override section you say:

> TODO Even better, mypy will warn you during static analysis.

Unfortunately, I can't express a type variable constraint based on another type variable so `@implements` can only expose itself as one of two things currently:

- `Callable[[KlassType], KlassType]`, so mypy does see that I'm not touching the underlying class.
- `Callable[[InterfaceType], InterfaceType]`, here mypy will correctly check that the implementation implements the interface, but it won't see that I'm returning the same class... It'll be treated as if I returned the `InterfaceType`...

In other languages I could have expressed that `KlassType` is a subclass of `InterfaceType` but well... I guess we'll have to wait. :)

**Reply by pauleveritt** [reply]
@Finistere Would you mind if I extracted your Injected Customer example into its own discussion post? I alas *almost* know what you're talking about, but not quite.

**Reply by Finistere** [reply]
Of course, feel free to ask any questions!

**Comment by pauleveritt** [comment]
Let me ask a bigger question: are the examples worth it?

They overlap with your quite-good tutorial. If you think it's worth it, I will write a bunch more.

**Reply by Finistere** [reply]
> quite-good tutorial

Thanks!

Your examples made me realize a few things: `@dataclass` integration, the `VisitHandler` to manipulate the `Visit` could be written in a better way, override mechanism... So I'm really grateful for the examples. However, integrating them into the documentation is another story as I'd need to merge it with the tutorial and maintain those examples. Work that I'm currently not willing to do. I'll definitely need to rework the tutorial later to integrate the new features though. At that time I'll re-visit your examples. :)

**Reply by pauleveritt** [reply]
Would you like me to manage the tutorial? Or do the merging?

**Reply by Finistere** [reply]
If you're interested in managing the tutorial, you're more than welcome! Before you start working on it, at the risk of stating the obvious, here is what I'm thriving for:
- the tutorial targets developers who never used Antidote before but want a step-by-step guide through it. As such it must show elementary usage of the most important features and link to recipes/how-to/... for advanced use/specific topics.
- Examples should be as concise as possible. Senior developers should be able to skim through the tutorial easily and junior ones shouldn't be lost.
- All code examples **must** be tested. I'm using `doctest` everywhere. It also often implies checking that the injection actually works, not just dependency declaration.

Ideally, I would also like examples to be type-checked but I've never set it up. I've started using `pyright` for recent tests but mypy didn't work properly. Obviously, the current documentation is not fault-free, but as said that is what I'm thriving for. :)

I'm using slightly changing the documentation in this PR https://github.com/Finistere/antidote/pull/49 which gets bigger and bigger, but I'm close to finishing it. ^^"

**Comment by pauleveritt** [comment]
I will also add, I'm having trouble getting `mypy` to recognize `py.typed` in Antidote for an editable install. I wonder if I've done the developer setup in the correct way.

**Reply by Finistere** [reply]
If you're working on Antidote itself I'm surprised it doesn't work, for `mypy` I'm just doing `mypy --strict src`. But I haven't used Antidote in other projects as an editable install. I'll try it out.

**Comment by pauleveritt** [comment]
I'm thinking that maybe your tutorial should stay as it is (and I can help on it) and perhaps my project should be thought of as a separate work. Almost like a book. Perhaps in a separate repo, but that's up to you.
